go/vt/vtgate/planbuilder/operators/horizon_planning.go

```diff
@@ -19,76 +19,179 @@ package operators
+			op, err := planDerived(ctx, in)
 			if err != nil {
 				return nil, false, err
 			}
 			return op, rewrite.NewTree, err
-		case *Filter:
-			err := planFilter(ctx, in)
-			if err != nil {
-				return nil, false, err
-			}
-			return in, rewrite.SameTree, nil
+		case *Projection:
+			return tryPushingDownProjection(ctx, in)
 		default:
 			return in, rewrite.SameTree, nil
 		}
 	}
 
-	newOp, err := rewrite.BottomUp(root, TableID, visitor, stopAtRoute)
+	newOp, err := rewrite.FixedPointBottomUp(root, TableID, visitor, stopAtRoute)
 	if err != nil {
 		if vterr, ok := err.(*vterrors.VitessError); ok && vterr.ID == "VT13001" {
 			// we encountered a bug. let's try to back out
-			return nil, errNotHorizonPlanned
+			return nil, errHorizonNotPlanned
 		}
 		return nil, err
 	}
 
 	return newOp, nil
 }
 
-func planHorizon(ctx *plancontext.PlanningContext, in *Horizon, isRoot bool) (ops.Operator, error) {
-	rb, isRoute := in.Source.(*Route)
-	if isRoot && !isRoute && ctx.SemTable.NotSingleRouteErr != nil {
-		// If we got here, we don't have a single shard plan
-		return nil, ctx.SemTable.NotSingleRouteErr
+func tryPushingDownProjection(
+	ctx *plancontext.PlanningContext,
+	p *Projection,
+) (ops.Operator, rewrite.ApplyResult, error) {
+	switch src := p.Source.(type) {
+	case *Route:
+		op, err := rewrite.Swap(p, src)
+		if err != nil {
+			return nil, false, err
+		}
+		return op, rewrite.NewTree, nil
+	case *ApplyJoin:
+		return pushDownProjectionInApplyJoin(ctx, p, src)
+	case *Vindex:
+		return pushDownProjectionInVindex(ctx, p, src)
+	default:
+		return p, rewrite.SameTree, nil
 	}
-	if isRoute && rb.IsSingleShard() && in.Select.GetLimit() == nil {
-		return planSingleRoute(rb, in)
+}
+
+func pushDownProjectionInVindex(
+	ctx *plancontext.PlanningContext,
+	p *Projection,
+	src *Vindex,
+) (ops.Operator, rewrite.ApplyResult, error) {
+	for _, column := range p.Columns {
+		expr := column.GetExpr()
+		_, _, err := src.AddColumn(ctx, aeWrap(expr))
+		if err != nil {
+			return nil, false, err
+		}
 	}
+	return src, rewrite.NewTree, nil
+}
+
+// pushDownProjectionInApplyJoin pushes down a projection operation into an ApplyJoin operation.
+// It processes each input column and creates new JoinColumns for the ApplyJoin operation based on
+// the input column's expression. It also creates new Projection operators for the left and right
+// children of the ApplyJoin operation, if needed.
+func pushDownProjectionInApplyJoin(ctx *plancontext.PlanningContext, p *Projection, src *ApplyJoin) (ops.Operator, rewrite.ApplyResult, error) {
+	var (
+		// idx is the current column index in the Projection's columns.
+		// in is the current column's ProjExpr.
+		idx int
+		in  ProjExpr
 
-	sel, isSel := in.Select.(*sqlparser.Select)
+		// lhsCols and rhsColumns store the columns to be added to the left and right child projections.
+		// lhsNames and rhsNames store the corresponding column names.
+		lhsCols, rhsColumns []ProjExpr
```

Comment:
nit: I think it would be `lhsCols` and `rhsCols` for both sides? unless there is some naming convention that i am missing for ApplyJoins

```suggestion
		// lhsCols and rhsCols store the columns to be added to the left and right child projections.
		// lhsNames and rhsNames store the corresponding column names.
		lhsCols, rhsCols []ProjExpr
```